### PR TITLE
Update signature for Deterministic Helper

### PR DESCRIPTION
**Pull request by tgroh**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`.
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---
This can be a protected, final method that specifically indicates that a
coder that is a component of this coder is not deterministic, and have a
more specific error message.

### CODE REVIEW

**Comment by tgroh** [comment]
This is a better change than #3029, which improves the name, and makes it more obvious how to use properly

**Comment by tgroh** [comment]
R: @lukecwik @kennknowles 
@davorbonaci

**Comment by tgroh** [comment]
None of these methods are used in the Dataflow Worker.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/Coder.java [hunk 1]
```diff
@@ -193,27 +193,21 @@ public T decode(InputStream inStream, Context context)
...
+    checkArgument(
+        this.equals(component),
+        "Cannot call verifyComponentDeterministic with the current %s %s as the component",
+        Coder.class.getSimpleName(),
+        this);
+    try {
+      component.verifyDeterministic();
+    } catch (NonDeterministicException e) {
+      throw new NonDeterministicException(this, message, e);
       }
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/Coder.java]: spacing here looks weird

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/Coder.java [hunk 2]
```diff
@@ -193,27 +193,21 @@ public T decode(InputStream inStream, Context context)
   public abstract void verifyDeterministic() throws Coder.NonDeterministicException;
 
   /**
-   * Verifies all of the provided coders are deterministic. If any are not, throws a {@link
+   * Verifies the provided coders is deterministic. If it is not, throws a {@link
    * NonDeterministicException} for the {@code target} {@link Coder}.
    */
-  public static void verifyDeterministic(Coder<?> target, String message, Iterable<Coder<?>> coders)
+  protected final void verifyComponentDeterministic(String message, Coder<?> component)
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/Coder.java]: Should this be on StructuredCoder?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/Coder.java (hunk 1 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/Coder.java]: Done.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/Coder.java (hunk 2 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/Coder.java]: We use it in at least one `CustomCoder`, as far as I know.

**Comment by tgroh** [comment]
This will not make the first stable release, so will require some minor changes. Primarily:

The static method will not be removed, but instead deprecated. All SDK callers will be ported away from that method onto the introduced "validateComponentDeterministic" method.

**Comment by davorbonaci** [comment]
@tgroh, is this really needed still? Perhaps the status quo is simpler.

**Comment by tgroh** [comment]
Closing as backwards incompatible. Filed https://issues.apache.org/jira/browse/BEAM-2579 for future work.

**Comment by tgroh** [comment]
Closing as backwards incompatible. Filed https://issues.apache.org/jira/browse/BEAM-2579 for future work.

**Pull request closed without merging**